Import Libraries

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Read in and Format Data

In [23]:
names = ['Age', 'Temperature of Water', 'Length of Fish']
df_x = pd.read_csv('x06Simple.csv', usecols=range(1,4), names=names, header=0, engine='python')
df_x = df_x.sample(frac=1, random_state=0)
df_y = pd.DataFrame(data=df_x[df_x.columns[-1]])
df_x.drop(df_x.columns[[-1,]], axis=1, inplace=True)

Split x and y Data into Test and Training Sets

In [16]:
n = np.int(np.ceil(df_x.shape[0] * 2 / 3))
df_x_train = df_x.head(n=n)
df_x_test = df_x.tail(n=df_y.shape[0] - n)
df_y_train = df_y.head(n=n)
df_y_test = df_y.tail(n=df_y.shape[0] - n)

Calculate Mean and Standard Deviation to Standardize x Training Data

In [17]:
mean = df_x_train.mean(axis=0)
std = df_x_train.std(axis=0)
df_x_s_train = df_x_train.subtract(mean).divide(std)
x = np.hstack([np.ones((df_x_s_train.shape[0], 1)), df_x_s_train.values])
y = df_y_train.values

Find Thetas

In [18]:
thetas = np.linalg.inv(x.T @ x) @ (x.T @ y)
print(thetas)

[[3275.66666667]
 [1097.60312694]
 [-259.32789051]]


As you can see, the resulting line is: $y = 3275.66666667 + 1097.60312694x_1 - 259.32789051x_2$

Define RMSE Function

In [21]:
def RMSE(thetas, x, y):
    x = np.hstack([np.ones((x.shape[0], 1)), x.subtract(mean).divide(std)])
    diff = y.subtract(x @ thetas)
    print(np.sqrt(diff.pow(2, axis=0).sum(axis=0).values/float(len(y))))

RMSE(thetas, df_x_test, df_y_test)

[601.9302895]


As you can see, the resulting RMSE is 601.9302895.